### ISSUE TITLE
Investigate question backup

### ISSUE THREAD

**Issue by Willem3141** [graphchecker/moodle-qtype_graphchecker#29]
In GitLab by @Willem3141 on Feb 18, 2020, 14:05

CodeRunner provided a class in the `backup` directory, that takes care of importing / exporting a CodeRunner question. We almost certainly broke this, so we should investigate and fix this.

**Comment by Willem3141** [comment]
In GitLab by @ArthurvG on Oct 20, 2020, 09:27

changed due date to December 17, 2020

**Comment by Willem3141** [comment]
In GitLab by @ArthurvG on Oct 22, 2020, 10:35

changed due date to December 10, 2020

**Comment by Willem3141** [comment]
In GitLab by @Willem3141 on Dec 14, 2020, 13:46

I just performed some testing with backup and restore. The good news is that I know how it works now, the bad news is that GraphChecker questions indeed do not work after backing up.

In fact, I had hoped that it would just work now that I removed the `backup` directory, considering that we are a "normal" question type. That is opposed to CodeRunner, which definitely needs special handling because it has an additional database table.

PHP warnings shown when trying to take a quiz with a restored GraphChecker question in it:

```
Failed to load question options from the table question_graphchecker_opts for questionid 175

Notice: Undefined property: stdClass::$answertype in /home/willem/Git/moodle/question/type/graphchecker/questiontype.php on line 107

Notice: Undefined property: stdClass::$answertype in /home/willem/Git/moodle/question/type/questiontypebase.php on line 923

Notice: Undefined property: stdClass::$answerpreload in /home/willem/Git/moodle/question/type/questiontypebase.php on line 923

[... dozens more of such warnings ...]
```

**Comment by Willem3141** [comment]
In GitLab by @Willem3141 on Dec 14, 2020, 13:54

Inspection with phpMyAdmin confirms that the corresponding rows in `question_graphchecker_opts` are not properly inserted when restoring. This is strange, since the documentation of `extra_question_fields()` in `questiontypebase.php` says that it will take care to "save, backup and restore the extra fields".

**Comment by Willem3141** [comment]
In GitLab by @Willem3141 on Feb 2, 2021, 13:10

removed due date

**Comment by Willem3141** [comment]
In GitLab by @Willem3141 on Feb 2, 2021, 13:10

changed due date to February 09, 2021

**Comment by Willem3141** [comment]
I tested exporting a question from the question bank, and importing it again, and it works fine now. I am not exactly sure what fixed the problem though?

**Comment by Willem3141** [comment]
It turns out that to backup the `extra_question_fields`, one still needs to inherit from `backup_qtype_extrafields_plugin` manually. We didn't do that (because we removed CodeRunner's code, being under the impression that the `extra_question_fields` would be backed up by Moodle automatically).

We found this out by looking at Moodle's built-in `shortanswer` question type, which does the same thing.
